File: tests-utils/src/main/java/tech/beshu/ror/utils/elasticsearch/IndexManager.java
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */
package tech.beshu.ror.utils.elasticsearch;

import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Sets;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpUriRequest;
import org.apache.http.util.EntityUtils;
import tech.beshu.ror.utils.httpclient.RestClient;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import static tech.beshu.ror.utils.misc.GsonHelper.deserializeJsonBody;

public class IndexManager {

  private final RestClient restClient;

  public IndexManager(RestClient restClient) {
    this.restClient = restClient;
  }

  public GetIndexResult get(String indexName) {
    try (CloseableHttpResponse response = restClient.execute(createGetIndexRequest(indexName))) {
      int statusCode = response.getStatusLine().getStatusCode();
      return statusCode != 200
          ? new GetIndexResult(statusCode, Sets.newHashSet())
          : new GetIndexResult(statusCode, getAliases(deserializeJsonBody(bodyFrom(response))));
    } catch (IOException e) {
      throw new IllegalStateException("Index manager get index result failed", e);
    }
  }


  private HttpUriRequest createGetIndexRequest(String indexName) {
    return new HttpGet(restClient.from("/" + indexName));
  }

  private static Set<String> getAliases(Map<String, Object> result) {
    List<Object> responses = result.values().stream().collect(Collectors.toList());
    return ((Map<String, Object>) ((Map<String, Object>)responses.get(0)).get("aliases")).keySet();
  }

  private static String bodyFrom(HttpResponse r) {
    try {
      return EntityUtils.toString(r.getEntity());
    } catch (IOException e) {
      throw new IllegalStateException("Cannot get string body", e);
    }
  }

  public static class GetIndexResult {

    private final Integer responseCode;
    private final ImmutableSet<String> aliases;

    GetIndexResult(Integer responseCode, Set<String> aliases) {
      this.responseCode = responseCode;
      this.aliases = ImmutableSet.copyOf(aliases);
    }

    public int getResponseCode() {
      return responseCode;
    }

    public ImmutableSet<String> getAliases() {
      return aliases;
    }
  }
}

File: tests-utils/src/main/scala/tech.beshu.ror.utils/elasticsearch/IndexManager.scala
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */
package tech.beshu.ror.utils.elasticsearch

import org.apache.http.client.methods.{HttpGet, HttpPost}
import org.apache.http.entity.StringEntity
import tech.beshu.ror.utils.elasticsearch.BaseManager.JsonResponse
import tech.beshu.ror.utils.httpclient.RestClient

class IndexManager(client: RestClient)
  extends BaseManager(client) {

  def getIndex(indices: String*): JsonResponse = {
    call(getIndexRequest(indices.toSet), new JsonResponse(_))
  }

  def getAliases: JsonResponse = {
    call(getAliasRequest(), new JsonResponse(_))
  }

  def getAlias(indices: String*): JsonResponse = {
    val indexOpt = indices.mkString(",") match {
      case "" => None
      case str => Some(str)
    }
    call(getAliasRequest(indexOpt), new JsonResponse(_))
  }

  def getAliasByName(index: String, alias: String): JsonResponse = {
    call(getAliasRequest(Some(index), Some(alias)), new JsonResponse(_))
  }

  def createAliasOf(index: String, alias: String): JsonResponse = {
    call(createAliasRequest(index, alias), new JsonResponse(_))
  }

  def createAliasAndAssert(index: String, alias: String): Unit = {
    val result = createAliasOf(index, alias)
    if(!result.isSuccess) {
      throw new IllegalStateException(s"Cannot create alias '$alias'; returned: ${result.body}")
    }
  }

  def getSettings(index: String*): JsonResponse = {
    call(createGetSettingsRequest(index.toSet), new JsonResponse(_))
  }

  def getAllSettings: JsonResponse = {
    call(createGetSettingsRequest(Set("_all")), new JsonResponse(_))
  }

  private def getAliasRequest(indexOpt: Option[String] = None,
                              aliasOpt: Option[String] = None) = {
    val path = indexOpt match {
      case Some(index) =>
        aliasOpt match {
          case Some(alias) => s"$index/_alias/$alias"
          case None => s"$index/_alias"
        }
      case None =>
        aliasOpt match {
          case Some(alias) => s"_alias/$alias"
          case None => "_alias"
        }
    }
    new HttpGet(client.from(path))
  }

  private def getIndexRequest(indices: Set[String]) = {
    new HttpGet(client.from(indices.mkString(",")))
  }

  private def createAliasRequest(index: String, alias: String) = {
    val request = new HttpPost(client.from("_aliases"))
    request.addHeader("Content-type", "application/json")
    request.setEntity(new StringEntity(
      s"""{"actions":[{"add":{"index":"$index","alias":"$alias"}}]}""".stripMargin))
    request
  }

  private def createGetSettingsRequest(indices: Set[String]) = {
    new HttpGet(client.from(s"${indices.mkString(",")}/_settings"))
  }

}
